pack/pack.go

```diff
@@ -0,0 +1,281 @@
+		return []byte(`"data"`), nil
+	case Tree:
+		return []byte(`"tree"`), nil
+	}
+
+	return nil, errors.New("unknown blob type")
+}
+
+func (t *BlobType) UnmarshalJSON(buf []byte) error {
+	switch string(buf) {
+	case `"data"`:
+		*t = Data
+	case `"tree"`:
+		*t = Tree
+	default:
+		return errors.New("unknown blob type")
+	}
+
+	return nil
+}
+
+// Blob is a blob within a pack.
+type Blob struct {
+	Type   BlobType
+	Length uint32
+	ID     backend.ID
+	Offset uint
+}
+
+// GetReader returns an io.Reader for the blob entry e.
+func (e Blob) GetReader(rd io.ReadSeeker) (io.Reader, error) {
+	// seek to the correct location
+	_, err := rd.Seek(int64(e.Offset), 0)
+	if err != nil {
+		return nil, err
+	}
+
+	return io.LimitReader(rd, int64(e.Length)), nil
+}
+
+// Packer is used to create a new Pack.
+type Packer struct {
+	blobs []Blob
+
+	bytes uint
+	k     *crypto.Key
+	wr    io.Writer
+	hw    *backend.HashingWriter
+
+	m sync.Mutex
+}
+
+// NewPacker returns a new Packer that can be used to pack blobs
+// together.
+func NewPacker(k *crypto.Key, w io.Writer) *Packer {
+	return &Packer{k: k, wr: w, hw: backend.NewHashingWriter(w, sha256.New())}
+}
+
+// Add saves the data read from rd as a new blob to the packer. Returned is the
+// number of bytes written to the pack.
+func (p *Packer) Add(t BlobType, id backend.ID, rd io.Reader) (int64, error) {
+	p.m.Lock()
+	defer p.m.Unlock()
+
+	c := Blob{Type: t, ID: id}
+
+	n, err := io.Copy(p.hw, rd)
+	c.Length = uint32(n)
+	c.Offset = p.bytes
+	p.bytes += uint(n)
+	p.blobs = append(p.blobs, c)
+
+	return n, err
+}
+
+var entrySize = binary.Size(BlobType(0)) + binary.Size(uint32(0)) + backend.IDSize
+
+// headerEntry is used with encoding/binary to read and write header entries
+type headerEntry struct {
+	Type   BlobType
+	Length uint32
+	ID     [backend.IDSize]byte
+}
+
+// Finalize writes the header for all added blobs and finalizes the pack.
+// Returned are the complete number of bytes written, including the header.
+// After Finalize() has finished, the ID of this pack can be obtained by
+// calling ID().
+func (p *Packer) Finalize() (int64, error) {
+	p.m.Lock()
+	defer p.m.Unlock()
+
+	// n tracks the number of bytes written for the header
+	var n int64 = int64(p.bytes)
+
+	// create writer to encrypt header
+	wr := crypto.EncryptTo(p.k, p.hw)
+
+	// write header
+	for _, b := range p.blobs {
+		entry := headerEntry{
+			Type:   b.Type,
+			Length: b.Length,
+		}
+		copy(entry.ID[:], b.ID)
+
+		err := binary.Write(wr, binary.LittleEndian, entry)
+		if err != nil {
+			return int64(n), err
+		}
+
+		n += int64(entrySize)
+	}
+
+	// finalize encrypted header
+	err := wr.Close()
+	if err != nil {
+		return int64(n), err
+	}
+
+	// account for crypto overhead
+	n += crypto.Extension
+
+	// write length
+	err = binary.Write(p.hw, binary.LittleEndian, uint32(len(p.blobs)*entrySize+crypto.Extension))
+	if err != nil {
+		return int64(n), err
+	}
+	n += int64(binary.Size(uint32(0)))
+
+	p.bytes = uint(n)
```

Comment:
just confused why we do any casting at all then... if everything will always fit into a `uint` (even when added up), why not use `uint` everywhere (instead of `uint64`)?